doc/barclamp.md

```diff
@@ -0,0 +1,1224 @@
+  this will allow you to deploy only your own barclamp's controller side
+  components to this node while all other OpenStack services reside on the main
+  controller. This greatly shortens the time a chef run for your barclamp's
+  controller side components takes and thus speeds up development. Make a
+  backup of this node as well to guard against breakage and allow for
+  clean-slate deployment of your barclamp's controller side components.
+
+The best way to set up your scratch node is to make it a lonely node. To this
+end, you will need to re-run mkcloud after you have created your cloud:
+
+```
+nodenumberlonelynode=1 mkcloud setuplonelynodes crowbar_register
+
+```
+
+## Development Environment
+
+Most of your development will take place on your Crowbar node (log in as root).
+You will find the currently running Crowbar code (both Rails application
+components and chef recipes) in the `/opt/dell` directory. We will provide
+detailed locations for all important components of a Barclamp in the _Basic
+Implementation_ section below.
+
+In order to track changes you make we recommend you keep a checkout of your
+`crowbar-openstack` fork on the Crowbar node and apply the changes you make in
+there
+
+### Preparations for Crowbar Node
+
+These are recommendations, but we'll assume you followed them for the remainder
+of this document since this shortens the commands in our examples. First of
+all, clone your `crowbar-openstack` fork to `/root/crowbar-openstack` on your
+Crowbar node and switch to your development branch.
+
+Next, create the following /root/.bashrc with a few useful aliases and
+environment variables (substitute with appropriate values where so indicated by
+comments):
+
+```
+# Replace "Crowbar Committer" by your name as it should appear in commit
+# messages.
+C_NAME="Crowbar Committer"
+
+# Replace "email@address.invalid" by your email address as it should appear in
+# commit messages.
+C_EMAIL="email@address.invalid"
+
+# Replace "6cefbd80254141a63f4c86df195cd930ac7eb10b" by your first commit's ID.
+export COMMIT="6cefbd80254141a63f4c86df195cd930ac7eb10b"
+
+# Insert your own barclamp's name here:
+export BARCLAMP_NAME=mybarclamp
+
+# Insert a space delimited list of the Barclamps you are working on here (only
+# required if your barclamp requires changes to other barclamps as well).
+
+export BARCLAMPS="$BARCLAMP_NAME"
+
+# Insert the path to your barclamp collection checkout here (in case it's
+# somewhere other than /root/crowbar-openstack)
+
+export BARCLAMP_FORK_PATH=/root/crowbar-openstack
+
+export GIT_AUTHOR_NAME="$C_NAME"
+export GIT_AUTHOR_EMAIL="$C_EMAIL"
+export GIT_COMMITTER_NAME="$C_NAME"
+export GIT_COMMITTER_EMAIL="$C_EMAIL"
+
+alias Cpatch="(cd $BARCLAMP_FORK_PATH; git show | patch -N -p1 --merge -d /opt/dell)"
+
+# Full patch: applies your fork completely against /opt/crowbar
+function Fpatch()
+  {
+  cd $BARCLAMP_FORK_PATH
+  git diff "$COMMIT^" HEAD | patch -N --merge -p1 -d /opt/dell
```

Comment:
When I create new files this is working for me only the first time. Will create files in /opt/dell.
But running it second time it's showing:
```
The next patch would create the file chef/cookbooks/monasca/README.md,
which already exists!  Skipping patch.
1 out of 1 hunk ignored
```
for all new files. So new files are not updated with new content.
It will patch only already existing files:
```
patching file chef/cookbooks/barbican/recipes/api.rb
patching file chef/cookbooks/horizon/recipes/server.rb
```